lib/Value/Matrix.pm

```diff
@@ -416,58 +451,105 @@ sub isZero {
 	return 1;
 }
 
-#
-#  See if the matrix is triangular, diagonal, symmetric, orthogonal
-#
+=head3 C<isUpperTriangular>
+
+Check if a Matrix is upper triangular (for degree > 2, applies to frontal slice matrices)
+
+=cut
 
 sub isUpperTriangular {
 	my $self = shift;
 	my @d    = $self->dimensions;
 	return 1 if scalar(@d) == 1;
-	return 0 if scalar(@d) > 2;
-	for my $i (2 .. $d[0]) {
-		for my $j (1 .. ($i - 1 < $d[1] ? $i - 1 : $d[1])) {
-			return 0 unless $self->element($i, $j) == 0;
+	if (scalar(@d) == 2) {
+		for my $i (2 .. $d[0]) {
+			for my $j (1 .. ($i - 1 < $d[1] ? $i - 1 : $d[1])) {
+				return 0 unless $self->element($i, $j) == 0;
+			}
+		}
+	} else {
+		for my $row (@{ $self->{data} }) {
+			if (!$row->isUpperTriangular) {
+				return 0;
+			}
```

Comment:
I would make this
```suggestion
			return 0 if !$row->isUpperTriangular;
```
Just a suggestion.  I like the more condensed form, and this is consistent with how these sort of checks are done elsewhere in the file.

Reply:
`unless` is certainly preferred.  Several of us have expressed preference for `statement unless condition` to `statement if !condition`.  If the condition is more complicated though, I prefer using `statement if complicated condition`.  `unless` can be confusing when you have a lot of conditions some of which are negated and others not with `&&`s and `||`s mixed in.